packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
-      final fileLocationsMap =
-          (json['locations'] as Map).cast<String, dynamic>();
-
-      for (final String file in fileLocationsMap.keys) {
-        final entries =
-            (fileLocationsMap[file] as Map).cast<String, List<dynamic>>();
-
-        final ids = entries['ids']!.cast<int>();
-        final lines = entries['lines']!.cast<int>();
-        final columns = entries['columns']!.cast<int>();
-        final names = entries['names']!.cast<String>();
-
-        for (var i = 0; i < ids.length; i++) {
-          final location = RebuildLocation(
-            id: ids[i],
-            path: file,
+  void incrementCount(Location location, int count) {
+    final id = location.id;
+    final existing = _rebuildLocationMap[id];
+    if (existing != null) {
+      existing.buildCount += count;
+    } else {
+      final rebuildLocation = RebuildLocation(location, buildCount: count);
+      _locations.add(rebuildLocation);
+      _rebuildLocationMap[id] = rebuildLocation;
+    }
+  }
+
+  void clear() {
+    _locations.clear();
+    _rebuildLocationMap.clear();
+  }
+
+  void dataChanged() {
+    _locations.notifyListeners();
+  }
+}
+
+/// Mapping from ids to [Location] objects.
+///
+/// Some Locations may be unresolved with ids but no valid path while valid
+/// paths are being fetched from the running Flutter application.
+class LocationMap {
+  LocationMap();
+
+  LocationMap.fromJson(Map<String, dynamic> json) {
+    processLocationMap(json);
+  }
+
+  final _locationMap = <int, Location>{};
+
+  int _countUnkownLocations = 0;
+
+  void clear() {
+    _countUnkownLocations = 0;
+    _locationMap.clear();
+  }
+
+  ValueListenable<bool> get locationsResolved => _locationsResolved;
+  final _locationsResolved = ValueNotifier(true);
+
+  Location operator [](int id) {
+    var location = _locationMap[id];
+    if (location == null) {
+      _countUnkownLocations++;
+      _locationsResolved.value = false;
+      // Add a placeholder location until we receive an event with the
+      // path + line + column for the location.
+      _locationMap[id] = location = Location(id: id);
+    }
+    return location;
+  }
+
+  Map<String, dynamic> toJson() {
+    final json = <String, dynamic>{};
+    final pathToLocations = <String, List<Location>>{};
+    for (var location in _locationMap.values) {
+      if (location.isResolved) {
+        pathToLocations
+            .putIfAbsent(location.path!, () => <Location>[])
+            .add(location);
+      }
+    }
+    pathToLocations.forEach((path, locations) {
+      final ids = <int>[];
+      final lines = <int>[];
+      final columns = <int>[];
+      final names = <String>[];
+      for (var location in locations) {
+        ids.add(location.id);
+        lines.add(location.line!);
+        columns.add(location.column!);
+        names.add(location.name!);
+      }
+      json[path] = {
+        'ids': ids,
+        'lines': lines,
+        'columns': columns,
+        'names': names,
```

Comment:
pull these strings out into consts to be used here and below

Reply:
done